### DISCUSSION TITLE
BRouter route duration

### DISCUSSION THREAD

**Discussion by tweaks90** [devemux86/cruiser#532] [Q&A] [answered]
Hi

I do have a little problem understanding the duration calculation in BRouter (offline.)

it seems that with version 4.0.0 cruiser only uses the average speed setting for the route duration calculation. Unfortunately, this can be completely off. I had an average speed setting of 45km/h, which does produce absolutely absurd durations on motorways. OTOH if I change the average to, e.g. 120 (which is more sensible on motorways), this would completely underestimate a route through smaller roads.

In 4.0.0 changing the setting to 'calculated route' is possible but gives the very same result as 'average speed'.

Any suggestions on how to get more usable duration estimations?

Thanks,
Tom

**Comment by MotoUKRider** [comment]
@tweaks90 I observe the same the same results here: Both average+calculated options produce the same route duration and the speed slider will vary the route duration, but - as you say - produce odd times if the roads are faster/slower than the speed you choose.

There is an issue with certain road types and them not having a single speed limit.  A motorway/trunk road isn't a problem, but when you come to Primary+Secondary roads, you need to consider some of these have lower speed limits as they pass through villages etc.  Some Trunk roads (certainly in the UK) have the same variations.

The `maxspeed` value is not always present in the OSM map data, so there are places where the routing engine cannot determine a speed from the road data - so it has to rely on an "average" value or possibly a "Speed vs Road Type" slider for each road.  The latter would be my preference - and provide the ability to tailor the calculations to your riding style and/or Routing Profile like: Fastest, Curvy, Sunday Touring or whatever.

**Comment by devemux86** [comment]
@tweaks90 thanks for the report.

BRouter currently uses the selected average speed to calculate the route duration.

We had discussed and made this change because BRouter `vmax` option also changes the route duration.
Some users select a lower `vmax` to calculate different routes, but then BRouter returns a longer duration.

If you all are comfortable with this, we can return in the next Cruiser update to the previous behavior:
we can also use with BRouter the duration calculated from the route.

The duration option still exists for those wanting to select route duration from average speed or calculated route.

**Reply by MotoUKRider** [reply]
@devemux86 Are you able to interject a user setting as per my above comment like "Primary Road Speed = 60" (or whatever value we set) or are we tied to whatever data is within BRouter?

**Reply by devemux86** [reply]
@MotoUKRider 

BRouter route duration is calculated by the router using the OpenStreetMap values combined with the selected `vmax`.
Everything that can be changed is done through the BRouter profiles. I don't know if we can insert such values there?

**Reply by MotoUKRider** [reply]
@devemux86 OK, so the following (from an older profile I had on my HD) *cannot* be changed in the Cruiser app?

`assign maxspeed_implicit =`
`switch highway=motorway                  999`
`switch highway=motorway_link             130`
`switch highway=trunk                     250`
`switch highway=trunk_link                100`
`switch highway=primary|primary_link      100`
`switch highway=secondary|secondary_link   90`
`switch highway=tertiary|tertiary_link     80`
`switch highway=unclassified               50`

These speeds are quite, ahem, generous!

**Reply by Ciutnik** [reply]
We already have an avoidance factor for individual road categories, set by the user. Maybe - if possible - you could also add a speed parameter to each of them?

**Reply by devemux86** [reply]
The speeds in the profiles are also used as a method to include / exclude many things.
Don't think that changing the speeds just affects the duration, it could made the profile unusable.

So everything requires to first understand how the profiles work and then many tests.
It is often better to not change the vmax and just play with the many avoid options.

In any case if you have a specific suggestion, please test it first on [BRouter website](https://brouter.de/brouter-web/).
This is the best way to test a profile and what the changes actually do to the calculated route.

**Reply by Ciutnik** [reply]
I meant defining speed only for calculating durations, not for calculating routes. But now I understand that arrival times are calculated by Brouter, not Cruiser. If Cruiser doesn't know the road categories on the route, my suggestion really doesn't make sense.

**Reply by devemux86** [reply]
- Route duration is calculated by BRouter with option "Calculated route".
In previous Cruiser, but we will return this behavior to avoid confusion.

- Route duration is calculated by Cruiser with option "Average speed".

---

Note that the `vmax` option is used internally in the BRouter code,
so we cannot affect it in the profiles.

**Reply by Ciutnik** [reply]
> Route duration is calculated by Cruiser with option "Average speed".

Ok I see. But Cruiser can't take into account different road categories because only BRouter knows them? Cruiser can only know the length of the entire route?

**Reply by devemux86** [reply]
Road info is available in the BRouter routes.

Using default speeds per road type, this is what all routers (and BRouter) already do.

Then users will ask extra settings to define the speed for all road types. :slightly_smiling_face: 
This can only be used for the duration of the route which is never very accurate.

If you mean this, we can think it for the future.
Can you please create a new discussion?

**Reply by MotoUKRider** [reply]
For info, adjusting the `assign maxspeed_implicit =` values in BRouter Web has a reasonably positive effect on route duration calculations :smile:  I plotted a route up the central part of the UK towards Scotland including some off-motorway parts, tweaking the values and got some better results.

Oddly, if I set them really high ~500 then it doesn't give me a really short duration. I've been unable to get an unrealistically short duration out of it. Also, I got a 2-minute difference when changing Motorway from 500 to 999.

As @devemux86 says, some more understanding required but initial tests allow for *some* tweaking but I can't ascertain as yet why it won't let me travel at supersonic speeds to test the values :shrug:

**Reply by tweaks90** [reply]
Getting the 'Calculatd Route' setting back sounds like a good idea.
Also, I think it could indeed make sense (but have no proof of course) to allow configuring average speeds per road type. Such that the 'Average speed' setting in Cruiser will use these values instead of the ones in BRouter. I guess a single average for all roads won't do here as there are simply too many different road types to cover it all.

**Reply by devemux86** [reply]
> Getting the 'Calculatd Route' setting back sounds like a good idea.

@tweaks90 it will be available again in the next Cruiser update.

**Comment by devemux86** [comment, marked as answer]
> Getting the 'Calculatd Route' setting back sounds like a good idea.

It is available again in [Cruiser 4.0.1](https://play.google.com/store/apps/details?id=gr.talent.cruiser).

**Comment by MotoUKRider** [comment]
Users switching from the various "Fast" / "Curvy" routing profiles may want to reset them, or at least investigate whether the calculated route duration has any relation to reality... I had a "fast" A-to-B route planned and noticed the very similar "curvy" version taking over 4hrs instead of a little over 1hr30m.  The mileage difference was negligible.

There speeds were all the same, and road avoidances did *not* change in the "curvy" profile when I reset it.  Unless I missed something very silly, I wonder if values from these profiles were being miscalculated in the recent version with the "calculated route" fix?

**Reply by bios4** [reply]
It heavily depends on the BRouter settings what finally is really calculated by the app...

I did NOT (yet) manage to get BRouter profiles set up in a way that
- "faster" is really "shorter overall duration" all the time
- "fast & curvy" really has a "shorter duration than 'curvy' alone"
- "fastest" produces itineraries with really the shortest duration

And because I do NOT want to spend too much time to finally figure all of this out during the riding season, I plan my routes with Kurviger for the time being. Maybe I invest more time for that off-season, but who knows... 😉

**Reply by devemux86** [reply]
@MotoUKRider Have you changed the BRouter speed option?  (default 160 km/h)

This `vmax` option is used by BRouter code to also calculate the route duration.
So changing this option (e.g. 50 km/h), it is expected to return a larger duration.

Or select the "Average speed" in Cruiser route duration setting
and set your own personal average speed to calculate the route duration.

**Reply by MotoUKRider** [reply]
@devemux86 They were all around 99, which makes it even more bizarre!  As the route pattern was only slightly different, the duration difference just didn't make any sense.  "Avoid high speed roads" were all around 99(mph).  I notice that the kmph/mph label for that slider only shows *both* units if you adjust it manually.  eg: it says 160km, but if I adjust it - I get the mph in (brackets).

**Reply by MotoUKRider** [reply]
@devemux86 The issue is "Prefer forests and green areas" - when ticked, it added a crazy amount of duration to the route despite little road alteration.  It may be that this feature is more for hiking/walking, so has a time penalty or something?

**Reply by devemux86** [reply]
> I notice that the kmph/mph label for that slider only shows both units if you adjust it manually. eg: it says 160km, but if I adjust it - I get the mph in (brackets).

@MotoUKRider thanks for the report, it will be fixed.

**Reply by devemux86** [reply]
> "Prefer forests and green areas"

These options are still developed in BRouter repository. And probably we need improvements in the driving profile.
Better to use the average speed with these options.

Also it is better to stay with the normal avoid options and do not use the options in the "More" screen. :slightly_smiling_face:

**Reply by MotoUKRider** [reply]
@devemux86 I don't remember ticking that option, but it took some trial'n'error before I found the cause. I looked on the BRouter web portal and noticed that "forest" / "green" are not mentioned in the "car" profile, but found a reference in the "Hiking" one, so as you say - one to avoid!

**Reply by devemux86** [reply]
They only added them in their bike and hiking profiles.

I then added them in the Cruiser driving profile. :slightly_smiling_face:
